src/main/java/org/elasticsearch/index/mapper/core/BinaryFieldMapper.java

```diff
@@ -144,10 +147,16 @@ public BytesReference value(Object value) {
         }
         try {
             if (indexCreatedBefore2x) {
-                return CompressorFactory.uncompressIfNeeded(bytes);
-            } else {
-                return bytes;
+                try {
+                    return CompressorFactory.uncompressIfNeeded(bytes);
+                } catch (NotXContentException e) {
+                    // This is a BUG! We try to decompress by detecting a header in
```

Comment:
Yes maybe add a warning, and add to the comment that the buggy behavior is kept for backwards compatibility?